Support C++11

Relates to https://github.com/mapbox/rasterio/issues/1230 and https://github.com/sgillies/frs-wheel-builds/issues/25

GDAL 2.3 will [require C++11](https://trac.osgeo.org/gdal/changeset/40984).  I don't think a Fiona build has been triggered on Travis since the change, but [Rasterio's GDAL trunk builds are failing with a bunch of warnings](https://travis-ci.org/mapbox/rasterio/jobs/315171560) like:

```
/home/travis/gdalinstall/gdal-trunk/include/ogr_geometry.h:974:9: warning: override controls (override/final) only available with -std=c++11 or -std=gnu++11 [enabled by default]
             const CPL_OVERRIDE;
             ^
```

The [Travis builds against GDAL trunk](https://travis-ci.org/Toblerity/Fiona/jobs/316454091) are currently failing with the same issues.

--- a/setup.py
+++ b/setup.py
@@ -180,6 +180,11 @@ ext_options = dict(
     libraries=libraries,
     extra_link_args=extra_link_args)
 
+ext_options_cpp = ext_options.copy()
+# GDAL 2.3+ requires C++11
+ext_options_cpp["extra_compile_args"] = ["-std=c++11"]
+
+
 # Define the extension modules.
 ext_modules = []
 
@@ -206,7 +211,7 @@ if source_is_repo and "clean" not in sys.argv:
 
     ext_modules = cythonize([
         Extension('fiona._geometry', ['fiona/_geometry.pyx'], **ext_options),
-        Extension('fiona._transform', ['fiona/_transform.pyx'], **ext_options),
+        Extension('fiona._transform', ['fiona/_transform.pyx'], **ext_options_cpp),
         Extension('fiona._crs', ['fiona/_crs.pyx'], **ext_options),
         Extension('fiona._drivers', ['fiona/_drivers.pyx'], **ext_options),
         Extension('fiona._err', ['fiona/_err.pyx'], **ext_options),
@@ -216,7 +221,7 @@ if source_is_repo and "clean" not in sys.argv:
 # If there's no manifest template, as in an sdist, we just specify .c files.
 elif "clean" not in sys.argv:
     ext_modules = [
-        Extension('fiona._transform', ['fiona/_transform.cpp'], **ext_options),
+        Extension('fiona._transform', ['fiona/_transform.cpp'], **ext_options_cpp),
         Extension('fiona._geometry', ['fiona/_geometry.c'], **ext_options),
         Extension('fiona._crs', ['fiona/_crs.c'], **ext_options),
         Extension('fiona._drivers', ['fiona/_drivers.c'], **ext_options),

--- a/setup.py
+++ b/setup.py
@@ -182,7 +182,10 @@ ext_options = dict(
 
 ext_options_cpp = ext_options.copy()
 # GDAL 2.3+ requires C++11
-ext_options_cpp["extra_compile_args"] = ["-std=c++11"]
+if sys.platform == "win32":
+    ext_options_cpp["extra_compile_args"] = ["/std:c++11"]
+else:
+    ext_options_cpp["extra_compile_args"] = ["-std=c++11"]
 
 
 # Define the extension modules.

